### ISSUE TITLE
DALL-E: Zero-Shot Text-to-Image Generation

### ISSUE THREAD

**Issue by yskim0** [One-Day-One-Paper/One-Day-One-Paper#21]
# Abstract

- text-to-image generation
- based on a transformer that autoregressively models the text and image tokens as a single stream of data
- also good at zero-shot fashion

# Introduction

- text to image
    - dataset → MS-COCO, CUB
- past approaches
    - multi-scale generators, integrating attention and auxiliary losses ...
- demonstrate that training a 12-billion parameter autoregressive transformer on 250 million image-text pairs
- DALL-E achieves high quality image generation on MS-COCO dataset *zero-shot*

# Method

- goal : to train a transformer to autoregressively model the text and image tokens as a single stream of data
- model should be spent capturing high-frequency details instead of the low-frequency structure that makes objects visually recognizable to us
- **using a two-stage training precedure**
    - **Stage 1. training a discrete variational autoencoder(dVAE)**
        - compress 256x256 RGB image into 32x32 grid of image tokens, each element of which can assume 8192 possible values
    - **Stage 2. train an autoregressive transformer to model the joint distribution over the text and image tokens**
        - 256 BPE(Byte Pair Encoding?) encoded text tokens with the 32x32=1024 image token
- overall procedure can be viewed as maximizing the evidencde lower bound(ELB) on the joint likelihood of the model distribution over images $x$, captions $y$, and the tokens $z$ for the encoded RGB image
    
    $\ln p_{\theta,\psi}(x,y) \geq \mathbb{E}_{z\sim q_\phi (z|x)} (\ln p_\theta(x|y,z) - \beta D_{KL}(q_\phi(y,z|x),p_\psi(y,z)))$
    
    - $q_\phi$ : distribution over the 32x32 image tokens generated by the dVAE encoder given the RGB image $x^2$
    `tokenize`
    - $p_\theta$ : distribution over the RGB image generated by the dVAE decoder given the image tokens
    `decode`
    - $p_\psi$ : the joint distribution over the text and image tokens modeled by the transformer

### Stage One : Learning the Visual Codebook

> Stage 1. training a discrete variational autoencoder(dVAE)
> 
> - compress 256x256 RGB image into 32x32 grid of image tokens, each element of which can assume 8192 possible values
- reduces the context size of the transformer by a factor of 192 without a large degradation in visual quality
    
    ![Untitled](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/b199aee3-e6db-4ffb-9ede-5b85223e0cdd/Untitled.png)
    
    위 : original image / botton : dVAE result 
    

- first stage's goal : maximize the ELB with respect to $\phi$  and $\theta$ , which corresponds to training a dVAE on the images alone
    
    > 이는 **이미지만에 대해 dVAE를 학습**한다는 뜻이다. ([https://littlefoxdiary.tistory.com/74](https://littlefoxdiary.tistory.com/74))
    > 
- We set the initial prior $p_\psi$ to the uniform categorical distribution over the K=8192 codebook vectors, and  **$q_\phi$ to be categorical distributions parametrized by the 8192 logits at the same spatial position in the 32x32 grid output by the encoder**
- ⇒ ELB becomes difficule to optimize
    - bacause, $q_\psi$ is a discrete distribution ⇒ we can't use the reparametrization gradient to maximize it
    - Alternative 1) online cluster assignment procedure + straight-through estimator
    - 💥 **Alternative 2) gumbel-softmax relaxation**
        - replacing the expectation over $q_\phi$ with one over $q_\phi^\tau$, where the relaxation becomes tight as the temperature $\tau \rightarrow 0.$
        - the likelihood for $p_\theta$ is evaluated using the log-laplace distirubiton
- the relaxed ELB is maximized using Adam with exponentially weighted iterate averaging
- for stable training
    - specific annealing schedules for the relaxation temperature and step size
        - annealing $\tau$ to $1/16$
        - (Appendix A.2.Training) 
        annealed from 1 to 1/16 over first 150,000 updates 
        Using a linear annealing scedule for this typically led to divergence
    - using 1x1 conv. at the end of the encoder and the beginning of the decoder
        - reducing the receptive field size for the ocnvolutions around the relaxation led to it generalizing better to the true ELB
        

### Code

```python
logits = self.encoder(img)
if return_logits:
	return logits # return logits for getting hard image indices for DALL-E training

temp = default(temp, self.temperature)
soft_one_hot = F.gumbel_softmax(logits, tau = temp, dim = 1, hard = self.straight_through)
sampled = einsum('b n h w, n d -> b d h w', soft_one_hot, self.codebook.weight)
out = self.decoder(sampled)
```
